feat(quantic): added jsdoc categories (#1566)
coveo/ui-kit: packages/quantic/docs/jsdoc-guidelines.md

> # JSDoc Guidelines for Quantic Components

Author: @jpmarceau what do you think of this?

Author: Could we have the component `categories` display as badges in the component's documentation?

Reviewer: I believe it would be possible to add badges, but I don't know if it's useful considering what I've implemented:
https://docsstaging.coveo.com/quantic-categories-demo/quantic/latest/reference/
The TOC simply reflects the category structure :) 
Do you think that's enough or we'd like a badge too?

Reviewer: BTW, I just realized by going through this Readme that components can be in multiple categories, and my script doesn't currently handle that, so you may see missing pages in my staging link until I fix that. (I'm missing pages for components that show up under multiple categories)

Author: Yeah i hadnt alerted you about that yet. That's the idea behind the badges. So we can support components being part of more than one category. Right now none of them are though.
The idea is not to limit ourselves with this new structure while still providing a way to view a relevant subset of components. Components with multiple categories would show up in multiple places. What do you think about that?

Reviewer: Yeah I can adjust my script so components show up in many categories when needed (that's the case of the ResultLink in fact).
I just don't really see the value of displaying badges if the components already show up in the target places in the TOC 🤔

Author: The idea behind the badges is 
1) to display all component metadata on the component page, which includes categories
2) so users aren't confused when they see it show up multiple places. I think the badges would make it more obvious that it's the same component

Reviewer: Ah yeah I see that makes sense 👍 
I'll implement something like what we do in headless for controllers that show up under several engines:
https://docs.coveo.com/en/headless/latest/reference/controllers/did-you-mean/#associated-engines

The Quantic library has [reference documentation](https://docs.coveo.com/en/quantic/latest/reference/) that is automatically generated from JSDoc comments in the component source code. The component comments are parsed by JSDoc and the resulting doclet data is parsed by our custom template script found in `quantic/docs/template/lwc-json/publish.js`. This document outlines the required attributes and format of these JSDOC comments.

> ## Type Imports

Reviewer: Just to make sure: it's normal that we don't leverage this in the documentation right now?

Author: Oh I hadn't noticed we didn't. I guess since there is no description of th event and when it is dispatched it's probably not important to include it yet... Frankly IDK.

Reviewer: Yeah I don't really see why someone would need that now 🤔 
Even in the future, I don't see why someone would need that info?

It is common to reference types from external dependencies, most notably Coveo Headless.
Types should be defined via `typedef` before the class definition.

example:
```js
/** @typedef {import("coveo").SearchEngine} SearchEngine */
```

## Components

> Components should be given a short but specific description explaining its role. Include relevant information about certain behaviours or limitations of the component here when pertinent to developers.

Reviewer: ```suggestion
Components should be given a short but specific description explaining their role. Include relevant information about certain behaviours or limitations of the component when pertinent to developers.
```

### @fires
Events fired from this component should be defined using the `@fires` tag.

### @category
Each component should be given one or more category values using the `@category` tag from one of the following cateogories:
- **Search**: Components pertaining to the Search use-case.
- **Case Assist**: Components pertaining to the Case Assist use-case.
- **Result Template**: Components meant to be used within result templates.
- **Utility**: Components providing a generic utility without pertaining to any specific Coveo Headless use-case.

> If there is more than one relevant category it should be specified with its own tag.

Reviewer: I'm not sure what this means. Do you mean that components could have multiple tags? If that's the case, here's a suggestion:
```suggestion
If there are more than one relevant category, each should be specified with its own tag.
```

If a new category is required it should be discussed with the code owners and adjustements must be made to the parsing script.

### @example
An example must be provided using the `@example` tag which includes inline values for all public properties. It is recommended to use values that are different from the default values of each property.

example:
```js
/**
 * The `QuanticFacetValue` component is used by a facet component to display a formatted facet value and the number of results with that value.
 * @fires CustomEvent#selectvalue
 * @category Search
 * @example
 * <c-quantic-facet-value onselectvalue={onSelect} item={result} is-checked={result.checked} display-as-link={displayAsLink} formatting-function={formattingFunction}></c-quantic-facet-value>
 */
export default class QuanticFacetValue extends LightningElement {
  ...
}
```

## Public Properties

> Public properties, marked using an `@api` **decorator** should be given a JSDoc comment following these guidelines.

Reviewer: ```suggestion
Public properties, marked using an `@api` **decorator**, should be specified via JSDoc comments following the guidelines below.
```

Each public property should be given a short but specific description explaining its role. Include relevant information about certain behaviours or limitations of the component here when pertinent to developers.

### @api

> Each public property should be given an `@api` tag in the JSDoc comment.

Reviewer: ```suggestion
Each public property should be specified with an `@api` tag in the JSDoc comment.
```

> Each public property should be specified with an `@api` tag in the JSDoc comment.

Reviewer: ```suggestion
Each public property documentation string should be preceded by an `@api` tag in the JSDoc comment.
```

### @type

> A type value must be specified for each public property. If the type is non-primitive as is the case if it is defined in Coveo Headless, a link to public documentation containing more detailed information should be provided.

Reviewer: ```suggestion
A type value must be specified for each public property. If the type is non-primitive as is the case when defined in Coveo Headless, a link to public documentation containing more detailed information should be provided.
```

- **Search**: Components pertaining to the Search use-case.
- **Case Assist**: Components pertaining to the Case Assist use-case.
- **Result Template**: Components meant to be used within result templates.
- **Utility**: Components providing a generic utility without pertaining to any specific Coveo Headless use-case.
If there is more than one relevant category it should be specified with its own tag.
If a new category is required it should be discussed with the code owners and adjustements must be made to the parsing script.

### @example
An example must be provided using the `@example` tag which includes inline values for all public properties. It is recommended to use values that are different from the default values of each property.

example:
```js
/**
 * The `QuanticFacetValue` component is used by a facet component to display a formatted facet value and the number of results with that value.
 * @fires CustomEvent#selectvalue
 * @category Search
 * @category Utility
 * @example
 * <c-quantic-facet-value onselectvalue={onSelect} item={result} is-checked={result.checked} display-as-link={displayAsLink} formatting-function={formattingFunction}></c-quantic-facet-value>
 */

> export default class QuanticBreadcrumbManager extends LightningElement {

Reviewer 2: :thinking: The example is wrong. It describes the `QuanticFacetValue` component but the exported component is the `QuanticBreadcrumbManager`...

Author: Lol I changed my mind and only copied the comment. Thanks for the catch!
